results2junit

```diff
@@ -0,0 +1,289 @@
+        if self.category:
+            xml += ">\n"
+            err_type = self.category.rstrip('s')
+            xml += "            <{} message=\"{}\">{}</{}>\n".format(
+                err_type,
+                self.message,
+                self.stacktrace,
+                err_type)
+            xml += "            <system-out>stdout</system-out>\n"
+            xml += "            <system-err>stderr</system-err>\n"
+            xml += "        </testcase>\n"
+        else:
+            xml += "/>\n"
+        return xml
+
+    @property
+    def stacktrace(self):
+        """
+        Returns the full docker-autotest stacktrace for a given testcase,
+        as read from the .ERROR file in this test's subdirectory.
+        """
+        # Given 'docker/subtests/docker_cli/run_volumes.46_756384709',
+        # read the file debug/run_volumes.46_756384709.ERROR in that subdir
+        testname = self.test_path.split("/")[-1]
+        error_log = os.path.join(self.test_path, 'debug', testname + '.ERROR')
+        stacktrace = ''
+        try:
+            with open(error_log, 'r') as error_log_fh:
+                stacktrace = error_log_fh.read()
+        except IOError, e:
+            stacktrace = e
+        return xml_escape(stacktrace)
+
+
+class AutotestResults(object):
+    """
+    Sample class for foo bar
+    """
+
+    def __init__(self):
+        status_file = 'status'
+        self.start_times = []
+        self.results = []
+        self.messages = ['']
+        with open(status_file, "r") as status_fh:
+            for line in status_fh:
+                line.strip()
+                parts = line.strip().rstrip().split("\t")
+                self.parse_status_line(parts)
+
+    def __iter__(self):
+        return (x for x in self.results)
+
+    def parse_status_line(self, parts):
+        """
+        Parses one autotest status line, eg
+            START/END GOOD   docker/this/that  ... timestamp ... message ...
+        Preserves all END lines inside self.results; START and status
+        lines have their timestamps and messages preserved for the
+        next END line.
+        """
+        if len(parts) < 5:
+            return
+        if not parts[3].startswith('timestamp='):
+            return
+        timestamp = int(parts[3].split("=")[1])
+
+        if parts[0].startswith('START'):
+            self.start_times.append(timestamp)
+            return
+
+        if parts[0].startswith('END'):
+            self.results.append({'name': parts[2],
+                                 'status': parts[0].replace('END', '').strip(),
+                                 'timestamp': timestamp,
+                                 'run_time': timestamp - self.start_times.pop(),
+                                 'result': self.messages.pop()})
+            return
+
+        if parts[5]:
+            self.messages.append(parts[5])
+
+    def pop(self):
+        """
+        Allows treating this obj as a list. Returns the last element.
```

Comment:
Because all I need is `pop`, and because if I subclass `list` I'm not sure how to store other state in `self`. I'm running into diminishing returns here, and don't want to spend a week polishing this up to perfection.